app/src/main/java/rick_and_morty/ui/characters/composables/CharacterList.kt

```diff
@@ -1,28 +1,54 @@
 import androidx.compose.foundation.layout.Arrangement
 import androidx.compose.foundation.lazy.LazyColumn
 import androidx.compose.foundation.lazy.items
-import androidx.compose.runtime.Composable
-import androidx.compose.runtime.LaunchedEffect
-import androidx.compose.runtime.collectAsState
+import androidx.compose.foundation.lazy.rememberLazyListState
+import androidx.compose.runtime.*
 import androidx.compose.ui.unit.dp
 import androidx.lifecycle.viewmodel.compose.viewModel
+import rick_and_morty.data.model.CharacterResultsDto
 import rick_and_morty.ui.characters.CharactersViewModel
+import rick_and_morty.ui.widgets.RickAndMortyErrorDialog
+import rick_and_morty.ui.widgets.CircularProgressBar
 
 @Composable
 fun CharacterList(charactersViewModel: CharactersViewModel = viewModel(modelClass = CharactersViewModel::class.java)) {
 
-    val characters = charactersViewModel.characters.collectAsState(emptyList())
-    LazyColumn (
-        verticalArrangement = Arrangement.spacedBy(5.dp)
-    ) {
-
-        items(characters.value) {characters ->
-            CharacterRow(characterResultsDto = characters)
-        }
-        item {
-            LaunchedEffect(true) {
+    val characters = charactersViewModel.characters.collectAsState().value
+    if (characters.isLoading)
+    {
+        CircularProgressBar()
+    }
+    when {
+        characters.isFailure -> RickAndMortyErrorDialog(characters.failureMessage.toString())
+        else -> {
+            val scrollListState = rememberLazyListState()
+            LazyColumn (
+                verticalArrangement = Arrangement.spacedBy(5.dp),
+                state = scrollListState
+            ) {
+                items(characters.isSuccess) { characters ->
+                    CharacterRow(characterResultsDto = characters)
+                }
+            }
+            val userIsAtBottom by remember{
+                derivedStateOf {
+                    val layoutInfo = scrollListState.layoutInfo
+                    val visibleItemsInfo = layoutInfo.visibleItemsInfo
+                    if (layoutInfo.totalItemsCount == 0) {
+                        false
+                    } else {
+                        val lastVisibleItem = visibleItemsInfo.last()
+                        val viewportHeight = layoutInfo.viewportEndOffset + layoutInfo.viewportStartOffset
+                        (lastVisibleItem.index + 1 == layoutInfo.totalItemsCount &&
+                                lastVisibleItem.offset + lastVisibleItem.size <= viewportHeight)
+                    }
+                }
+            }
+            if (userIsAtBottom){
+                LaunchedEffect(true) {
                     charactersViewModel.getCharacters()
-          }
```

Comment:
Hi, sorry for the delay in reviewing this pull request. In #8,  multiple calls of `getCharacters` is fixed when scrolling. Have you tested if the same bug is fixed when orientation changes?  I guess `scrollListState` is probably lost when orientation changes because views are destroyed during orientation changes.

Since `UIState` is introduced in this pr, I think you can use this `isLoading` flag for preventing multipole calls of  `getCharacters`, because UIState survives during orientation changes and should re-emit the latest value after orientation changes.

Reply:
@pkliang yep, it is just calls ones when you scroll to till the end, and it doesn't call again when orientation changes.